hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/oauth2/AzureADAuthenticator.java

```diff
@@ -168,12 +168,14 @@ public static AzureADToken getTokenFromMsi(final String authEndpoint,
    */
   public static AzureADToken getTokenUsingRefreshToken(
       final String authEndpoint, final String clientId,
-      final String refreshToken) throws IOException {
+      final String clientSecret, final String refreshToken) throws IOException {
     QueryParams qp = new QueryParams();
     qp.add("grant_type", "refresh_token");
     qp.add("refresh_token", refreshToken);
-    if (clientId != null) {
```

Comment:
Nope, the client id is not optional in the refresh token grant flow. Thats a miss interpretation on your part of the spec.

I think you have been miss leading by how it's written in the spec in section 6
https://www.rfc-editor.org/rfc/rfc6749#section-6

here it doesn't say in that section that the client id is required, just because there is different way to inject the client credential into the request (ie POST or BASIC, even client assertion).
But you see in the example that the authorisation header is setup and this contains the client ID in there.

For confidential client, its quite obvious that the client ID is required. For public client (which you want to support too), it's also required, as said in:

```
authenticate the client if client authentication is included and
      ensure that the refresh token was issued to the authenticated
      client, and
```

Your code currrently only support one type of auth method, the POST one.
You then can assume that in this line, the client ID is required.

You can revisit injecting it in the body of the request if one day you decide to support the auth method BASIC as well. Would just mean the client ID is in the request but in a header instead.

Reply:
well, microsoft wrote that bit of the code. we can assume that they knew what their oauth endpoint does, though i personally find that endpoint's response to auth failures (200 + text/html message for humans) fairly suboptimal

Reply:
If it's microsoft that wrote this line, then why injecting the client ID if present? if it works without it...

As the fix is not about this particular line, I will just revert it so we concentrate on the real issue.